Handle operators when rewriting expression trees

333fred opened the pull request:
Currently, we're not handling binary operators in the MethodToClass rewriter, which causes user-defined operators to be missed when we lift things to closures. Fixes https://github.com/dotnet/roslyn/issues/56376. Recommend reviewing each commit separately: commit 1 nullable-enables MethodToClassRewriter, and commit 2 actually fixes the bug.

333fred commented:
@dotnet/roslyn-compiler for reviews.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs:
```diff
@@ -6002,5 +6002,241 @@ .maxstack  3
+        {
+            Func<Expression<Func<T, Wrapper<T>, bool>>> func = () => (x, y) => x == y;
+
+            func();
```
>func(); [](http://example.com/codeflow?start=12&length=7)

I think we should evaluate the expression tree and observe that it calls the operator. This is applicable to other tests.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs:
```diff
@@ -6002,5 +6002,241 @@ .maxstack  3
+  IL_0039:  call       ""System.Linq.Expressions.Expression<System.Func<BadImageFormatExceptionRepro.Program.Wrapper<T>, BadImageFormatExceptionRepro.Program.Wrapper<T>>> System.Linq.Expressions.Expression.Lambda<System.Func<BadImageFormatExceptionRepro.Program.Wrapper<T>, BadImageFormatExceptionRepro.Program.Wrapper<T>>>(System.Linq.Expressions.Expression, params System.Linq.Expressions.Paramete ...
+  IL_003e:  ret
+}");
+        }
```
>} [](http://example.com/codeflow?start=8&length=1)

Please ensure we have similate test coverage for VB

AlekseyTs replied:
>} [](http://example.com/codeflow?start=8&length=1)

It feels like we should also test user defined conversions and user defined short-circuiting operators (represented by BoundUserDefinedConditionalLogicalOperator nodes after initial binding)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs:
```diff
@@ -6002,5 +6002,241 @@ .maxstack  3
+  IL_0039:  call       ""System.Linq.Expressions.Expression<System.Func<BadImageFormatExceptionRepro.Program.Wrapper<T>, BadImageFormatExceptionRepro.Program.Wrapper<T>>> System.Linq.Expressions.Expression.Lambda<System.Func<BadImageFormatExceptionRepro.Program.Wrapper<T>, BadImageFormatExceptionRepro.Program.Wrapper<T>>>(System.Linq.Expressions.Expression, params System.Linq.Expressions.Paramete ...
+  IL_003e:  ret
+}");
+        }
```
>} [](http://example.com/codeflow?start=8&length=1)

The PR description indicates that there was a problem specifically around lifting things to closure, but it appears that none of the added tests cause lifting to occur. I guess that is fine if the fix can be observed without lifting, but it would be good to have at least one test per distinct scenario using it.

333fred replied:
It was really about the lowering to a separate class, not the closure fields themselves. I don't think there's anything to test with the actual closure hoist itself.

AlekseyTs commented:
Done with review pass (commit 2)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [925109942](https://github.com/dotnet/roslyn/pull/56598#issuecomment-925109942) [](http://example.com/codeflow?ancestors=925109942,925109942)

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/MethodToClassRewriter.cs:
```diff
@@ -518,16 +542,17 @@ public override BoundNode VisitFunctionPointerLoad(BoundFunctionPointerLoad node
+            var whenNotNull = (BoundExpression)this.Visit(node.WhenNotNull);
+            var whenNullOpt = (BoundExpression?)this.Visit(node.WhenNullOpt);
             TypeSymbol type = this.VisitType(node.Type);
             return node.Update(receiver, VisitMethodSymbol(node.HasValueMethodOpt), whenNotNull, whenNullOpt, node.Id, type);
```
>Debug.Assert(method is not null); [](http://example.com/codeflow?start=12&length=33)

Not sure what is the motivation for this change.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/MethodToClassRewriter.cs:
```diff
@@ -632,10 +633,10 @@ public override BoundNode VisitReadOnlySpanFromArray(BoundReadOnlySpanFromArray
-            return originalReceiver != null && originalReceiver.Kind == BoundKind.BaseReference &&
-                   rewrittenReceiver != null && rewrittenReceiver.Kind != BoundKind.BaseReference;
+            return originalReceiver is { Kind: BoundKind.BaseReference } &&
+                   rewrittenReceiver is { Kind: BoundKind.BaseReference };
```
>rewrittenReceiver is { Kind: BoundKind.BaseReference }; [](http://example.com/codeflow?start=19&length=55)

Consider not bundling refactorings like this with product changes in the same PR. This Also doesn't look like a change related to enabling nullable annotations. It is a good thing we have tests that cought this bug.

AlekseyTs reviewed (approved):
LGTM (commit 7)

333fred commented:
I'm going to rebase this PR now that Aleksey has reviewed it for ease of the next person to review it to separate out the nullable enable and the code change.

333fred commented:
@dotnet/roslyn-compiler for a second review.

333fred commented:
@dotnet/roslyn-compiler for a second review.

jcouv reviewed (approved):
LGTM Thanks (iteration 2). Don't forget to retarget the PR for 17.0
